File: app/src/main/java/com/example/zombiechat/chat/data/repo/FirebaseChatRepo.java
package com.example.zombiechat.chat.data.repo;

import android.util.Log;

import com.example.zombiechat.account.data.models.UserModel;
import com.example.zombiechat.chat.data.models.ChatRoomModel;
import com.example.zombiechat.chat.data.models.LastChatModel;
import com.example.zombiechat.chat.data.models.SingleChatModel;
import com.example.zombiechat.constants.api.collections.Collections;
import com.example.zombiechat.constants.fields.Fields;
import com.google.android.gms.tasks.Task;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import io.reactivex.rxjava3.core.Observable;

public class FirebaseChatRepo implements ChatRepo {

    final String currentUser = Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getUid();
    final CollectionReference chatRoomCollection = FirebaseFirestore.getInstance().collection(Collections.chatRoomIds);
    final CollectionReference chatCollections = FirebaseFirestore.getInstance().collection(Collections.chatCollection);
    final CollectionReference friendsCollection = FirebaseFirestore.getInstance().collection(Collections.friendsCollection);
    final CollectionReference usersCollection = FirebaseFirestore.getInstance().collection(Collections.userCollection);

    @Override
    public Observable<List<LastChatModel>> getLastChats() throws ExecutionException, InterruptedException {
        List<LastChatModel> lastChatModels = new ArrayList<>();

        return Observable.create(emitter -> {

            // Creating a task which will complete in future with charRoomIds
            Task<QuerySnapshot> chatRoomTask = chatRoomCollection.document(currentUser).collection(Collections.chatRoomIds).get();

            // Attaching success listener for now.
            chatRoomTask.addOnSuccessListener(charRoomsSnapshot -> {


                for (DocumentSnapshot chatRoom : charRoomsSnapshot.getDocuments()) {

                    // converting chatModel
                    ChatRoomModel chatRoomModel = Objects.requireNonNull(chatRoom.toObject(ChatRoomModel.class));

                    // task for chats in a chat room
                    Task<QuerySnapshot> chatsTaskForAChatRoom = chatCollections.document(chatRoomModel.getChatRoomId()).collection(Collections.chatCollection).orderBy(Fields.timestamp, Query.Direction.DESCENDING).limit(1).get();

                    // Attaching a success listener
                    chatsTaskForAChatRoom


                            .addOnSuccessListener(chats -> {
                                for (DocumentSnapshot chat : chats.getDocuments()) {

                                    // task for user details to show in chat list with last message
                                    Task<DocumentSnapshot> userTask = usersCollection.document(chatRoomModel.getFriendId()).get();
                                    SingleChatModel singleChatModel = chat.toObject(SingleChatModel.class);

                                    // Attaching a success listener
                                    userTask.addOnSuccessListener(user -> {

                                        UserModel userModel = user.toObject(UserModel.class);
                                        LastChatModel lastChatModel = new LastChatModel(chatRoomModel.getChatRoomId(), Objects.requireNonNull(singleChatModel).getMessage(), Objects.requireNonNull(userModel).getImage(), userModel.getName(), userModel.getUserid(), Objects.requireNonNull(singleChatModel).getTime()

                                        );

                                        // Adding last chat model to list
                                        lastChatModels.add(lastChatModel);
                                        emitter.onNext(lastChatModels);
                                    });


                                }


                            });
                }


            });


        });


    }


    @Override
    public void sendMessage(String message, String sendTo, String chatRoomId) {
        SingleChatModel singleChatModel = new SingleChatModel();
        singleChatModel.setMessage(message);
        singleChatModel.setSendBy(currentUser);
        singleChatModel.setSentTo(sendTo);
        singleChatModel.setTime(Timestamp.now());
        chatCollections.document(chatRoomId).collection(Collections.chatCollection).add(singleChatModel);
    }
}

File: app/src/main/java/com/example/zombiechat/src/chat/data/repo/FirebaseChatRepo.kt
package com.example.zombiechat.src.chat.data.repo

import android.util.Log
import com.example.zombiechat.src.account.data.models.UserModel
import com.example.zombiechat.src.chat.data.models.ChatRoomModel
import com.example.zombiechat.src.chat.data.models.LastChatModel
import com.example.zombiechat.src.chat.data.models.SingleChatModel
import com.example.zombiechat.util.consts.DbCollection
import com.example.zombiechat.util.consts.Fields
import com.example.zombiechat.util.service.ChatException
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import java.util.Objects
import java.util.concurrent.ExecutionException

class FirebaseChatRepo : ChatRepo {
    val currentUser: String = FirebaseAuth.getInstance().currentUser?.uid ?: ""
    val chatRoomCollection: CollectionReference =
        FirebaseFirestore.getInstance().collection(DbCollection.chatRoomIds)
    val chatCollections: CollectionReference =
        FirebaseFirestore.getInstance().collection(DbCollection.chatCollection)
    val friendsCollection: CollectionReference =
        FirebaseFirestore.getInstance().collection(DbCollection.friendsCollection)
    val usersCollection: CollectionReference =
        FirebaseFirestore.getInstance().collection(DbCollection.userCollection)

    override fun lastChats(): Observable<List<LastChatModel>> {
        val lastChatModels: MutableList<LastChatModel> = ArrayList()

        // TODO : fix the order of list view in streams.
        return Observable.create { emitter: ObservableEmitter<List<LastChatModel>> ->

            // Creating a task which will complete in future with charRoomIds
            val chatRoomTask =
                chatRoomCollection.document(currentUser).collection(DbCollection.chatRoomIds)
                    .get()


            // Attaching success listener for now.
            chatRoomTask.addOnSuccessListener { charRoomsSnapshot: QuerySnapshot ->
                for (chatRoom in charRoomsSnapshot.documents) {
                    // converting chatModel

                    val chatRoomModel = chatRoom.toObject(
                        ChatRoomModel::class.java
                    ) ?: continue

                    // task for chats in a chat room
                    val chatsTaskForAChatRoom =
                        chatCollections.document(chatRoomModel.chatRoomId!!)
                            .collection(DbCollection.chatCollection).orderBy(
                                Fields.timestamp, Query.Direction.DESCENDING
                            ).limit(1).get()

                    // Attaching a success listener
                    chatsTaskForAChatRoom


                        .addOnSuccessListener { chats: QuerySnapshot ->
                            for (chat in chats.documents) {
                                // task for user details to show in chat list with last message

                                val userTask =
                                    usersCollection.document(chatRoomModel.friendId!!).get()
                                val singleChatModel = chat.toObject(
                                    SingleChatModel::class.java
                                )

                                // Attaching a success listener
                                userTask.addOnSuccessListener { user: DocumentSnapshot ->
                                    val userModel = user.toObject(
                                        UserModel::class.java
                                    )
                                    if (singleChatModel != null && userModel != null) {
                                        val lastChatModel = LastChatModel(
                                            chatRoomModel.chatRoomId!!,
                                            singleChatModel.message ?: "",
                                            userModel.image ?: "",
                                            userModel.name ?: "",
                                            userModel.userid ?: "",
                                            singleChatModel.time ?: Timestamp.now()

                                        )

                                        // Adding last chat model to list
                                        lastChatModels.add(lastChatModel)
                                        lastChatModels.sortWith { o1: LastChatModel, o2: LastChatModel ->
                                            o2.msgTime.compareTo(
                                                o1.msgTime
                                            )
                                        }
                                        emitter.onNext(lastChatModels)
                                    }
                                }
                            }
                        }
                }
            }
        }
    }

    override fun getActiveChats(chatRoomId: String?): Observable<List<SingleChatModel>> {
        val singleChatModels: MutableList<SingleChatModel> = ArrayList()
        return Observable.create { emitter: ObservableEmitter<List<SingleChatModel>> ->
            chatCollections.document(
                chatRoomId!!
            ).collection(DbCollection.chatCollection)
                .orderBy(Fields.timestamp, Query.Direction.ASCENDING)
                .addSnapshotListener { value: QuerySnapshot?, error: FirebaseFirestoreException? ->
                    if (error != null) {
                        Log.e("ChatRepo", "Error in getting chat messages", error)
                    }
                    if (value != null) {
                        for (chat in value.documents) {
                            val singleChatModel = chat.toObject(
                                SingleChatModel::class.java
                            )
                            if (singleChatModel != null) {
                                singleChatModels.add(singleChatModel)
                            }
                        }
                        emitter.onNext(singleChatModels)
                    }
                }
        }
    }


    override fun sendMessage(message: String?, sendTo: String?, chatRoomId: String?) {

        try {
            val singleChatModel = SingleChatModel()
            singleChatModel.message = message
            singleChatModel.sendBy = currentUser
            singleChatModel.sentTo = sendTo
            singleChatModel.time = Timestamp.now()
            chatCollections.document(chatRoomId!!).collection(DbCollection.chatCollection)
                .add(singleChatModel)
        } catch (e: Exception) {
            throw ChatException("Error in sending message", e)
        }

    }
}
